### PR TITLE
Make public API setters fluent

### PR DESCRIPTION
**Pull request by felixbarny**
depends on elastic/apm-agent-java#293

### CODE REVIEW

**Comment by ngourley** [comment]
This is pretty awesome. To support method chaining in 0.7 I was creating extension methods in Kotlin.

**Comment by felixbarny** [comment]
Ideally, we should also remove the `set` prefix but so far I don't think it justifies a breaking change or deprecating the `set` methods and introducing others without the prefix.

**Comment by ngourley** [comment]
Any update on this?

**Comment by felixbarny** [comment]
It needs some changes to be compatible with the old API which has void return types. After the changes it can be merged :)

**Comment by felixbarny** [comment]
thanks for the ping btw

**Comment by felixbarny** [comment]
I have confirmed that this also works when using version `1.1.0` of the API

File: apm-agent-api/src/main/java/co/elastic/apm/api/AbstractSpanImpl.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        Object span = doCreateSpan();
+        return span != null ? new SpanImpl(span) : NoopSpan.INSTANCE;
+    }
+
+    private Object doCreateSpan() {
+        // co.elastic.apm.plugin.api.AbstractSpanInstrumentation$DoCreateSpanInstrumentation.doCreateSpan
+        return null;
+    }
+
+    void doSetName(String name) {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-api/src/main/java/co/elastic/apm/api/AbstractSpanImpl.java]: I don't think so, as package private is more restrictive. IntelliJ suggests removing the `protected` qualifier.

File: apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/LegacySpanInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+    private final ElementMatcher<? super MethodDescription> methodMatcher;
+
+    public LegacySpanInstrumentation(ElementMatcher<? super MethodDescription> methodMatcher) {
+        this.methodMatcher = methodMatcher;
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return named("co.elastic.apm.api.SpanImpl").and(not(hasSuperType(named("co.elastic.apm.api.AbstractSpanImpl"))));
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/LegacySpanInstrumentation.java]: Nice

**Pull request merged**